crates/flux-infer/src/refine_tree.rs

```diff
@@ -780,3 +791,328 @@ impl RefineCtxtTrace {
+    }
+}
+
+#[derive(Debug)]
+struct Graph {
+    /// description of each constraint
+    edges: FxHashMap<CstrId, CstrInfo>,
+    /// set of edges where kvid appears as ASSM
+    kv_lhs: FxHashMap<KVid, FxHashSet<CstrId>>,
+    /// set of edges where kvid appears as HEAD
+    kv_rhs: FxHashMap<KVid, FxHashSet<CstrId>>,
+}
+
+impl Graph {
+    fn new(node: &Node) -> Self {
+        let mut graph = Self {
+            edges: FxHashMap::default(),
+            kv_lhs: FxHashMap::default(),
+            kv_rhs: FxHashMap::default(),
+        };
+        graph.build(node, &mut vec![]);
+        graph
+    }
+
+    fn head(&self, cid: CstrId) -> VertexId {
+        let info = self
+            .edges
+            .get(&cid)
+            .unwrap_or_else(|| tracked_span_bug!("missing constraint info for {:?}", cid));
+        info.rhs
+    }
+
+    fn cstrs_with_kvar_as_assumption(&self, kvid: &KVid) -> Vec<CstrId> {
+        self.kv_lhs
+            .get(kvid)
+            .unwrap_or(&FxHashSet::default())
+            .iter()
+            .copied()
+            .collect()
+    }
+
+    fn build(&mut self, node: &Node, ctx: &mut Vec<KVid>) {
+        let n = ctx.len();
+        match &node.kind {
+            NodeKind::Head(expr, _) => {
+                let mut vertices = vec![];
+                expr_vertices(expr, &mut vertices);
+                for rhs in &vertices {
+                    self.insert_edge(ctx, *rhs);
+                }
+            }
+            NodeKind::Assumption(expr) => {
+                let mut vertices = vec![];
+                expr_vertices(expr, &mut vertices);
+                for v in &vertices {
+                    if let VertexId::KVar(kvid) = v {
+                        ctx.push(*kvid);
+                    }
+                }
+            }
+            _ => {}
+        };
+
+        for child in &node.children {
+            self.build(&child.borrow(), ctx);
+        }
+
+        ctx.truncate(n); // restore ctx
+    }
+
+    fn insert_edge(&mut self, lhs: &[KVid], rhs: VertexId) {
+        let edge_id = CstrId(self.edges.len());
+
+        // Create and insert edge
+        let edge_info = CstrInfo { lhs: lhs.iter().copied().collect(), rhs };
+        self.edges.insert(edge_id, edge_info);
+
+        // Add edge_id to kv_lhs for each kvar in lhs
+        for kvid in lhs {
+            self.kv_lhs.entry(*kvid).or_default().insert(edge_id);
+        }
+        // Add edge_id to kv_rhs for rhs kvar
+        if let VertexId::KVar(kvid) = rhs {
+            self.kv_rhs.entry(kvid).or_default().insert(edge_id);
+        }
+    }
+
+    fn kvars(&self) -> impl Iterator<Item = KVid> + '_ {
+        self.kv_lhs
+            .keys()
+            .chain(self.kv_rhs.keys())
+            .copied()
+            .collect::<FxHashSet<_>>()
+            .into_iter()
+    }
+}
+
+/// helper function to split an `Expr` into its constituent `VertexId`s
+fn expr_vertices(expr: &Expr, vertices: &mut Vec<VertexId>) {
```

Comment:
We implement something similar inside Expr::flatten_conj. I would factor that out and expose it. Something like

```rust
        pub fn visit_conj(&self, f: &mut impl FnMut(&Expr)) {
            if let ExprKind::BinaryOp(BinOp::And, e1, e2) = self.kind() {
                e1.visit_conj(f);
                e2.visit_conj(f);
            } else {
                f(self)
            }
        }
```

Then we can use this method in the two places where you are calling expr_vertices and avoid the intermediate allocation.

Reply:
You are collecting into a vec and then immediately iterating over that vec and discarding it.

Reply:
I rejigged this, take a look to see if its what you meant?